From 188deea607b17eb041e0a3505f5b9e17352acd1d Mon Sep 17 00:00:00 2001
From: Sujeet Baranwal <sbaranwal@marvell.com>
Date: Tue, 19 Nov 2019 13:39:22 -0800
Subject: [PATCH 0852/1921] proc: Board info logic reorg

Code has been reshuffled to make board info reside in static
memory than parsing DTB on demand. The DTB read would now happen
only once and the cached info would be printed upon requests.

Bug: IPBUCE-914

Change-Id: Idf19358a7700bbe26e73df1bbd6b2369ba33c8a0
Signed-off-by: Sujeet Baranwal <sbaranwal@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/18973
Reviewed-by: Sujeet Kumar Baranwal <Sujeet.Baranwal@cavium.com>
Reviewed-by: Derek Chickles <Derek.Chickles@cavium.com>
Reviewed-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
Tested-by: sa_ip-sw-jenkins <sa_ip-sw-jenkins@marvell.com>
Reviewed-on: https://sj1git1.cavium.com/c/IP/SW/kernel/linux/+/32743
Reviewed-by: Chandrakala Chavva <cchavva@marvell.com>
Tested-by: Sunil Kovvuri Goutham <Sunil.Goutham@cavium.com>
Signed-off-by: Sunil Goutham <sgoutham@marvell.com>
[WK: The original patch got from Marvell sdk11.21.09]
Signed-off-by: Wenlin Kang <wenlin.kang@windriver.com>
---
 drivers/soc/marvell/octeontx_info.c | 90 +++++++++++++++++++++--------
 1 file changed, 65 insertions(+), 25 deletions(-)

diff --git a/drivers/soc/marvell/octeontx_info.c b/drivers/soc/marvell/octeontx_info.c
index cbeebdb70b01..33b3926a3eb3 100644
--- a/drivers/soc/marvell/octeontx_info.c
+++ b/drivers/soc/marvell/octeontx_info.c
@@ -16,15 +16,27 @@
 #define MAC_ADDR_STR_LEN 18
 #define OCTTX_NODE "octeontx_brd"
 
+struct octtx_brd_info {
+	const char *board_revision;
+	const char *board_serial;
+	const char *board_model;
+	char board_mac[18];
+	int  dev_tree_parsed;
+};
+
 static struct proc_dir_entry *ent;
+static struct octtx_brd_info brd;
+
+static char null_string[5] = "NULL";
+static char expansion[MAC_ADDR_STR_LEN] = "00:00:00:00:00:00";
 
 static void format_ethaddr(const char *macPtr, char *out)
 {
-	char expansion[MAC_ADDR_STR_LEN] = "00:00:00:00:00:00";
 	int r = 0, w = 0, len;
 
 	/* Ignore 0x in the beginning */
-	macPtr = macPtr + 2;
+	if (!strncmp(macPtr, "0x", 2))
+		macPtr += 2;
 
 	len = strlen(macPtr);
 
@@ -44,30 +56,15 @@ static void format_ethaddr(const char *macPtr, char *out)
 
 static int oct_brd_proc_show(struct seq_file *seq, void *v)
 {
-	struct device_node *np = NULL;
-	const char *board_revision;
-	const char *board_serial;
-	const char *board_model;
-	const char *board_mac;
-	char mac_addr[18];
-
-	np = of_find_node_by_name(NULL, OCTTX_NODE);
-	if (!np) {
+	if (!brd.dev_tree_parsed) {
 		seq_puts(seq, "No board info available!\n");
 		return 0;
 	}
 
-	of_property_read_string(np, "BOARD-MODEL", &board_model);
-	of_property_read_string(np, "BOARD-REVISION", &board_revision);
-	of_property_read_string(np, "BOARD-SERIAL",  &board_serial);
-	of_property_read_string(np, "BOARD-MAC-ADDRESS", &board_mac);
-
-	format_ethaddr(board_mac, mac_addr);
-
-	seq_printf(seq, "Board model: %s\n", board_model);
-	seq_printf(seq, "Board revision: %s\n", board_revision);
-	seq_printf(seq, "Board serial: %s\n", board_serial);
-	seq_printf(seq, "Board MAC: %s\n", mac_addr);
+	seq_printf(seq, "board_model: %s\n", brd.board_model);
+	seq_printf(seq, "board_revision: %s\n", brd.board_revision);
+	seq_printf(seq, "board_serial_number: %s\n", brd.board_serial);
+	seq_printf(seq, "mac_addr_base: %s\n", brd.board_mac);
 
 	return  0;
 }
@@ -89,11 +86,54 @@ static const struct file_operations oct_brd_fops = {
 
 static int octtx_info_init(void)
 {
+	int ret;
+	const char *board_mac;
 	struct device_node *np = NULL;
 
-	np = of_find_node_by_name(NULL, OCTTX_NODE);
-	if (np)
-		ent = proc_create("octtx_info", 0660, NULL, &oct_brd_fops);
+	if (!brd.dev_tree_parsed) {
+		np = of_find_node_by_name(NULL, OCTTX_NODE);
+		if (!np) {
+			pr_err("No board info available!\n");
+			return 0;
+		}
+		ret = of_property_read_string(np, "BOARD-MODEL",
+						&brd.board_model);
+		if (ret) {
+			pr_err("Board model not available\n");
+			/* Default name is "NULL" */
+			brd.board_model = null_string;
+		}
+		ret = of_property_read_string(np, "BOARD-REVISION",
+						&brd.board_revision);
+		if (ret) {
+			pr_err("Board revision not available\n");
+			/* Default name is "NULL" */
+			brd.board_revision = null_string;
+
+		}
+		ret = of_property_read_string(np, "BOARD-SERIAL",
+						&brd.board_serial);
+		if (ret) {
+			pr_err("Board serial not available\n");
+			/* Default name is "NULL" */
+			brd.board_serial = null_string;
+		}
+		ret = of_property_read_string(np, "BOARD-MAC-ADDRESS",
+								&board_mac);
+		if (ret) {
+			pr_err("Board mac address not available\n");
+			/* Default name is "NULL" */
+			strncpy(brd.board_mac, expansion,
+						sizeof(brd.board_mac));
+		} else
+			format_ethaddr(board_mac, brd.board_mac);
+
+		brd.dev_tree_parsed = 1;
+	}
+
+	ent = proc_create("octtx_info", 0660, NULL, &oct_brd_fops);
+	if (ent == NULL)
+		return -ENODEV;
 
 	return 0;
 }
-- 
2.31.1

